lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -37,196 +35,206 @@ define((require, exports, module) ->
       return
 
-    # Bind a shortcut to tryPopup callback.
+    # Bind newPopup to Alt-Enter editor shortcut
     editor.commands.addCommand({
-      name: "test",
+      name: "newPopup",
       bindKey: "Alt-Enter",
-      exec: tryPopup
+      exec: newPopup
     })
-
     return
 
+
+  # Get the word under the cursor.
+  # @param {Editor} editor: editor object
+  # @return {String}
+  extractWord = (editor) ->
+    session = editor.session
+    row = editor.getCursorPosition().row
+    wordRange = getCurrentWordRange(editor)
+    # getAWordRange returns start and end positions with a trailing
+    # whitespace at the end (if there's a one), that's why replace is used
+    return session.getTextRange(wordRange).replace(/\s\s*$/, '')
+
+  # Returns Range object that describes current word position.
+  # @param {Editor} editor: editor object
+  # @return {Range}: current word range.
+  getCurrentWordRange = (editor) ->
+    session = editor.session
+    row = editor.getCursorPosition().row
+    col = editor.getCursorPosition().column
+    return session.getAWordRange(row, col)
+
+  # Convert an array of string to popup eligible structure.
+  # @param {Array} corrections: array of strings with substitution options.
+  # @return {Array}: array of jsons, actually.
+  convertCorrectionList = (corrections) ->
+    options = []
+    options.push({caption: item, value: item}) for item in corrections
+    return options
+
+
+  # TODO: incorrect token without correction list.
+  # Object contains init, show and detach functions for popup and some
+  # routines for inner use.
+  PopupManager =
+    # Bindings for editor keys with PopupManager functions.
+    commands:
+      "Esc": ->
+        PopupManager.detach()
+      "Up": ->
+        PopupManager.goTo("up")
+      "Down": ->
+        PopupManager.goTo("down")
+      "Return": ->
+        PopupManager.insertCorrection()
+
+    # Initial setup for popup.
+    # @param {Editor} editor: editor object.
+    init: (editor) ->
+      @editor = editor
+
+      HashHandler = require("ace/keyboard/hash_handler").HashHandler
+      AcePopup = require("ace/autocomplete/popup").AcePopup
+
+      @popup = new AcePopup(document.body)
+      @popup.setTheme(editor.getTheme())
+      @popup.setFontSize(editor.getFontSize())
+      @popup.on("click", (e) =>
+        @insertCorrection()
+        e.stop()
+      )
+
+      @session = editor.session
+
+      @keyboard = new HashHandler()
+      @keyboard.bindKeys(@commands)
+
+      @detach if editor?
+      return
+
+    # Show popup in editor.
+    # @param {Array} options: list of corrections for the current word.
+    show: (options) ->
+      @options = options
+
+      @popup.setData(options)
+
+      @editor.keyBinding.addKeyboardHandler(@keyboard)
+
+      position = @editor.getCursorPosition()
+      @base = @session.doc.createAnchor(position.row, position.column)
+
+      PopupManager.detach if editor?
+      @editor.on("change", @changeListener)
```

Comment:
we usually name such methods like `onChange`, `onMouseWheel`. Btw, one more way to get good `this` in event handlers is not to pass them directly, but wrap into a closure: `@editor.on("change", -> PopupManager.changeListener())` or even `self.changeListener()` if you save `this` to `self` in the beginning of this method.